templates/go/client.go.twig

```diff
@@ -40,74 +102,249 @@ func (clt *Client) Set{{header.key | caseUcfirst}}(value string) {
+		// Set Custom headers
+		for key, val := range headers {
+			req.Header.Set(key, toString(val))
+		}
+
+		if isGet {
+			q := req.URL.Query()
+			for key, val := range params {
+				rt := reflect.TypeOf(val)
+				switch rt.Kind() {
+				case reflect.Array:
+				case reflect.Slice:
+					arr := reflect.ValueOf(val)
+					for i := 0; i < arr.Len(); i++ {
+						q.Add(fmt.Sprintf("%s[]", key), toString(arr.Index(i)))
+					}
+				default:
+					q.Add(key, toString(val))
+				}
+			}
+			rawQuery := q.Encode()
+			req.URL.RawQuery = rawQuery
+		}
 	}
 
 	// Set Client headers
 	for key, val := range clt.headers {
-		req.Header.Set(key, ToString(val))
+		req.Header.Set(key, toString(val))
 	}
 
-	//Set default headers
-{% for key,header in spec.global.defaultHeaders %}
-	req.Header.Set("{{key}}", "{{header}}")
-{% endfor %}
-
-	// Set Custom headers
-	for key, val := range headers {
-		req.Header.Set(key, ToString(val))
+	// Make request
+	resp, err := clt.client.Do(req)
+	if err != nil {
+		return nil, err
 	}
 
-	if isGet {
-		q := req.URL.Query()
-		for key, val := range params {
-			q.Add(key, ToString(val))
-		}
-		req.URL.RawQuery = q.Encode()
-	}
+	// Handle response
+	defer resp.Body.Close()
 
-	// Make request
-	response, err := clt.client.Do(req)
+	responseData, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
 		return nil, err
 	}
 
-	// Handle response
-	defer response.Body.Close()
- 
-	responseData, err := ioutil.ReadAll(response.Body)
+	var isJson = !strings.HasPrefix(resp.Header.Get("content-type"), "text")
+	if isJson {
+		var jsonResponse map[string]interface{}
+		json.Unmarshal(responseData, &jsonResponse)
+		if resp.StatusCode < 200 || resp.StatusCode > 300 {
+			message, ok := jsonResponse["message"].(string)
+			if !ok {
+				message = "N/A"
+			}
+			return nil, &ClientError{
+				statusCode: resp.StatusCode,
+				message:    message,
+			}
+		}
+		jsonResponse["status"] = resp.StatusCode
+		jsonResponse["headers"] = resp.Header
+		return jsonResponse, nil
+	}
+	output, err := getOutput(responseData, resp.Header.Get("content-type"))
 	if err != nil {
 		return nil, err
 	}
+	if resp.StatusCode < 200 || resp.StatusCode > 300 {
+		return nil, &ClientError{
+			statusCode: resp.StatusCode,
+			message:    output,
+		}
+	}
+	return map[string]interface{}{
+		"status":  resp.StatusCode,
+		"headers": resp.Header,
+		"result":  output}, nil
+}
 
-	var jsonResponse map[string]interface{}
-	json.Unmarshal(responseData, &jsonResponse)
+func getOutput(responseData []byte, contentType string) (string, error) {
+	if strings.HasPrefix(contentType, "text") || strings.HasSuffix(contentType, "json") {
+		return string(responseData), nil
+	}
+	tmpFile, err := ioutil.TempFile(os.TempDir(), "prefix-")
+	if err != nil {
+		return "", err
+	}
+	if _, err = tmpFile.Write(responseData); err != nil {
+		return "", err
+	}
 
-	return jsonResponse, nil
+	if err := tmpFile.Close(); err != nil {
+		return "", err
+	}
+	return tmpFile.Name(), nil
```

Comment:
This should handle somehow larger, binary files.
It is still somehow hacky in that matter.  The idea was, instead of sending the (byte as text) output back, save the file to a tmp file and just return the file path.
The best thing would be a reader reference, as a member of the map.

Reply:
I am now using the `file` param to store the content of a binary file. So it matchtes with the upload behaviour.